### PR TITLE
{ACS} `az aks approuting`: Creating commands for approuting enable|disable|keyvault as well as dns zone add|remove|list

### PR DESCRIPTION
**Pull request by bfoley13**
**Related command**
az aks

**Description**
The app routing add-on has migrated to the stable api for AKS. This PR adds the same features within the preview cli ti the az aks cli.

This adds a flag to cluster create for enabling the addon:
az aks create ... --enable-app-routing

This also adds the approuting sub command to az aks:
az aks approuting [enable|disable|update]

And the zone sub command to the approuting command:
az aks approuting zone [add|delete|update|list]

**Testing Guide**

```
# Create a keyvault
az keyvault create --resource-group={resource_group} --location={location} --name={kv_name} --enable-rbac-authorization=false

# Create new AKS cluster enabling the app routing addon
aks create --resource-group={resource_group} --name={aks_name} --location={location}  --ssh-key-value={ssh_key_value} --enable-app-routing

# Use the approuting update command to link the keyvault
aks approuting update --resource-group={resource_group} --name={aks_name} --enable-kv --attach-kv {keyvault_id}

# Create a DNS Zone
network dns zone create --resource-group={resource_group} --name {dns_zone_1}

# Use the approuting zone command to add dns zone
aks approuting zone add --resource-group={resource_group} --name={aks_name} --ids {dns_zone_id_1} --attach-zones
```

**History Notes**

[ACS] `az aks create`: Add flag `--enable-app-routing` to enable app routing
[ACS] `az aks approuting`: Add command group to handle enable|disable|update of the approuting addon
[ACS] `az aks approuting zone`: Add command group to handle add|delete|update|list actions of dns zone resources associated to the approuting addon

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
ACS

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_abort.yaml [hunk 1]
```diff
@@ -13,10 +13,9 @@ interactions:
       ParameterSetName:
       - --resource-group --name --ssh-key-value --no-wait
       User-Agent:
-      - AZURECLI/2.49.0 azsdk-python-azure-mgmt-containerservice/24.0.0 Python/3.8.10
-        (Linux-5.15.0-1039-azure-x86_64-with-glibc2.29)
+      - AZURECLI/2.57.0 azsdk-python-core/1.28.0 Python/3.11.5 (macOS-14.3-x86_64-i386-64bit)
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_abort.yaml:16]: probably you don't need to update this file

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_api_server_authorized_ip_ranges.yaml [hunk 2]
```diff
@@ -14,10 +14,9 @@ interactions:
       - --resource-group --name --location --dns-name-prefix --node-count --ssh-key-value
         --load-balancer-sku --api-server-authorized-ip-ranges
       User-Agent:
-      - AZURECLI/2.49.0 azsdk-python-azure-mgmt-containerservice/24.0.0 Python/3.8.10
-        (Linux-5.15.0-1039-azure-x86_64-with-glibc2.29)
+      - AZURECLI/2.57.0 azsdk-python-core/1.28.0 Python/3.11.5 (macOS-14.3-x86_64-i386-64bit)
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_api_server_authorized_ip_ranges.yaml:17]: probably you don't need to update this file

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_approuting_update_with_monitoring_addon_enabled.yaml [hunk 3]
```diff
@@ -0,0 +1,2432 @@
+interactions:
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_approuting_update_with_monitoring_addon_enabled.yaml:1]: As you've mark the corresponding test case as live_only(), it's not necessary to commit this recording file

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py [hunk 4]
```diff
@@ -10153,3 +10153,570 @@ def test_aks_upgrade_upgrade_settings(self, resource_group, resource_group_locat
...
+                self.check("ingressProfile.webAppRouting.enabled", True),
+                self.check("addonProfiles.azureKeyvaultSecretsProvider.enabled", True),
+            ],
+        )
+
+        # delete cluster
+        delete_cmd = "aks delete --resource-group={resource_group} --name={aks_name} --yes --no-wait"
+        self.cmd(delete_cmd, checks=[self.is_empty()])
+
+    @live_only()
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py:10321]: please leave a comment explaining why it is marked as @live_only()

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py [hunk 5]
```diff
@@ -10153,3 +10153,570 @@ def test_aks_upgrade_upgrade_settings(self, resource_group, resource_group_locat
         # delete
         self.cmd(
             'aks delete -g {resource_group} -n {name} --yes --no-wait', checks=[self.is_empty()])
+
+    @AllowLargeResponse()
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py:10157]: Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=88311518&view=results) to validate the change.

- test_aks_create_with_app_routing_enabled
- test_aks_approuting_enable_disable
- test_aks_approuting_enable_with_keyvault_secrets_provider_addon_and_keyvault_id
- test_aks_approuting_update
- test_aks_approuting_update_with_monitoring_addon_enabled
- test_aks_approuting_zone_add_delete_list
- test_aks_approuting_zone_update

**Review by FumingZhang — COMMENTED** [review]
Please fix failed CI checks.

This PR moves the implementation in aks-preview to official azure-cli
- https://github.com/Azure/azure-cli-extensions/pull/6958
- https://github.com/Azure/azure-cli-extensions/pull/6968
- https://github.com/Azure/azure-cli-extensions/pull/7038
- https://github.com/Azure/azure-cli-extensions/pull/7197
- https://github.com/Azure/azure-cli-extensions/pull/7284

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_approuting_update_with_monitoring_addon_enabled.yaml (hunk 3 above)
Comment (bfoley13) [reply src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_approuting_update_with_monitoring_addon_enabled.yaml:1]: removed

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py (hunk 4 above)
Comment (bfoley13) [reply src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py:10321]: This one doesnt need it, removed!

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_api_server_authorized_ip_ranges.yaml (hunk 2 above)
Comment (bfoley13) [reply src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_api_server_authorized_ip_ranges.yaml:17]: removed these changes

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py (hunk 5 above)
Comment (bfoley13) [reply src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py:10157]: Created [new run with latest changes](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=88337759&view=results)

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_abort.yaml (hunk 1 above)
Comment (bfoley13) [reply src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_abort.yaml]: removed these changes

**Comment by FumingZhang** [comment]
Full test CI failed due to the recording of test_aks_approuting_enable_with_keyvault_secrets_provider_addon_and_keyvault_id is not valid, [reran the test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=88404292&view=results), may find the recording file from pipeline artifact and commit the new one.

File: src/azure-cli/azure/cli/command_modules/acs/_params.py [hunk 6]
```diff
@@ -717,6 +718,25 @@ def load_arguments(self, _):
...
+    with self.argument_context("aks approuting zone add") as c:
+        c.argument("dns_zone_resource_ids", options_list=["--ids"], required=True)
+        c.argument("attach_zones")
+
+    with self.argument_context("aks approuting zone delete") as c:
+        c.argument("dns_zone_resource_ids", options_list=["--ids"], required=True)
+
+    with self.argument_context("aks approuting zone update") as c:
+        c.argument("dns_zone_resource_ids", options_list=["--ids"], required=True)
+        c.argument("attach_zones")
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_params.py:738]: After some investigation, I found out that the reason why Check CLI Linter fails is due to the use of double quotes here. Please use single quote instead to pass CI.

Created an issue to the tool [#434](https://github.com/Azure/azure-cli-dev-tools/issues/434).

Comment (bfoley13) [reply src/azure-cli/azure/cli/command_modules/acs/_params.py]: updated to single quote

File: src/azure-cli/azure/cli/command_modules/acs/commands.py [hunk 7]
```diff
@@ -233,3 +233,20 @@ def load_command_table(self, _):
...
+        g.custom_command("update", "aks_approuting_update")
+
+    # AKS approuting dns-zone commands
+    with self.command_group(
+        "aks approuting zone", managed_clusters_sdk, client_factory=cf_managed_clusters
+    ) as g:
+        g.custom_command("add", "aks_approuting_zone_add")
+        g.custom_command("delete", "aks_approuting_zone_delete", confirmation=True)
+        g.custom_command("update", "aks_approuting_zone_update")
+        g.custom_command("list", "aks_approuting_zone_list")
```
[the remark this answers is not in the document]
Comment (bfoley13) [reply src/azure-cli/azure/cli/command_modules/acs/commands.py]: I really appreciate the help @FumingZhang! Updated to single quote

[the remark this answers is not in the document]
Comment (bfoley13) [reply src/azure-cli/azure/cli/command_modules/acs/commands.py]: still not fixed, seems like its complaining about mesh, but thats outside our changes

File: src/azure-cli/azure/cli/command_modules/acs/commands.py [hunk 8]
```diff
@@ -233,3 +233,20 @@ def load_command_table(self, _):
             'rollback',
             'aks_mesh_upgrade_rollback',
             supports_no_wait=True)
+
+    # AKS approuting commands
+    with self.command_group(
+        'aks approuting', managed_clusters_sdk, client_factory=cf_managed_clusters
+    ) as g:
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/commands.py:240]:
```suggestion
    with self.command_group('aks approuting', managed_clusters_sdk, client_factory=cf_managed_clusters) as g:
```

File: src/azure-cli/azure/cli/command_modules/acs/commands.py [hunk 9]
```diff
@@ -233,3 +233,20 @@ def load_command_table(self, _):
...
+        g.custom_command('enable', 'aks_approuting_enable')
+        g.custom_command('disable', 'aks_approuting_disable', confirmation=True)
+        g.custom_command('update', 'aks_approuting_update')
+
+    # AKS approuting dns-zone commands
+    with self.command_group(
+        'aks approuting zone', managed_clusters_sdk, client_factory=cf_managed_clusters
+    ) as g:
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/commands.py:248]:
```suggestion
    with self.command_group('aks approuting zone', managed_clusters_sdk, client_factory=cf_managed_clusters) as g:
```

Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/commands.py:248]: Another issue with the line break

Created issue [#435](https://github.com/Azure/azure-cli-dev-tools/issues/435)

**Review by FumingZhang — COMMENTED** [review]
Please also resolve merge conflict

File: src/azure-cli/azure/cli/command_modules/acs/commands.py (hunk 9 above)
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/commands.py:249]: don't forget this

Comment (bfoley13) [reply src/azure-cli/azure/cli/command_modules/acs/commands.py:249]: Totally missed this was a suggestion, pulling it in!

Comment (bfoley13) [reply src/azure-cli/azure/cli/command_modules/acs/commands.py]: added

File: src/azure-cli/azure/cli/command_modules/acs/commands.py (hunk 8 above)
Comment (bfoley13) [reply src/azure-cli/azure/cli/command_modules/acs/commands.py]: removed the line breaks in the command definitions

**Review by FumingZhang — APPROVED** [review]
LGTM

**Comment by yanzhudd** [comment]
Since it seems to have mutiply changes in this PR, it is better to split the changelog into several small items and add them in the History Notes part in the above description.
image

**Review by yanzhudd — APPROVED** [review]

**Pull request merged**